packages/front-end/services/experiments.ts

```diff
@@ -245,6 +245,25 @@ export function applyMetricOverrides(
   return { newMetric, overrideFields };
 }
 
+export function isExpectedDirection(
+  stats: SnapshotMetric,
+  metric: MetricInterface
+): boolean {
+  const expected: number = stats?.expected ?? 0;
+  return metric.inverse ? expected < 0 : expected > 0;
+}
+
+export function isStatSig(
+  stats: SnapshotMetric,
+  pValueThreshold: number
+): boolean {
+  const pValue: number = stats?.pValue ?? 1;
+  return pValue < pValueThreshold;
+}
+
 export function pValueFormatter(pValue: number) {
+  if (pValue === undefined) {
+    return "P-value missing";
+  }
```

Comment:
Another way to do this if we call this function in a lot of places and prefer that it itself is more defensive (rather than the implementer being required to do the checks), and if we want to keep UI strings out of utils (which will be slightly easier to localize if we decide to support other languages), we can do something like this:

```ts
export function pValueFormatter(pValue: number): string { // or : string | null, depending on style preference
  if (typeof pValue !== 'number') {
    return '' // or null, depending on style preference
  }

  return pValue < 0.001 ? "<0.001" : pValue.toFixed(3);
}
```

And the implementation:

```tsx
const pValueText = pValueFormatter(pValue) || "P-value missing";

return (
  <>{pValueText}</>
)
```

Again, this is mostly a stylistic preference. What you have is fine to fix the bug.

Reply:
I definitely prefer this solution, but now in order to share the fallback text across both components that call this formatter, I'd have to create yet another shared function somewhere in the front end, or duplicate the logic and string in two places. Do you think that's an issue or do you think it's worth it to get the string out of the util?

Reply:
There are multiple ways to approach this, and I think arguably they each have their merits, e.g.:

- keep it as is to reduce code duplication, and knowing that it will always be formatted this way, it's not a concern if there's a designer that wants it formatted a different way (we don't have a designer right now and it seems like you are leading what p-values look like so it seems fine to take this approach). 
- have the function return null (or empty string) and handle it in the UI with a fallback, e.g. `pValueFormatter(pValue) || 'P-value missing'`. This keeps the strings out of the UI but requires us to specify a fallback value every time it's used. I wouldn't create another function to avoid duplication as you say, I think since it's the UI it's ok if each implementation specifies it separately. This is personally the approach I would take but I mostly build localized apps—in Canada, it's a legal requirement for larger companies based in Canada to offer both English and French so I tend to always think in terms of accommodating localization. 

At this point it's like splitting hairs since we don't have those localization requirements. Priority-wise, I think it's important to get the fix in (especially since it looks like it could crash the UI) rather than try to find the perfect architecture for this reusable empty state string. In the grand scheme of things, if we do localize and forget about it because it's buried in a util, users of another language will see this string in English, which isn't the end of the world. The risk of making the wrong architectural decision in this case is next to nothing.

I already approved a while back. I will leave any style changes up to you.

Reply:
> I think since it's the UI it's ok if each implementation specifies it separately.

This seems like a good argument, so i'm gonna go with something like your second approach now.

> I already approved a while back. I will leave any style changes up to you.

Yep, there's definitely hair splitting here but I learn a lot from people like you with more experience working on these kinds of things, so thanks a bunch for the detailed feedback! Getting it is why I kept asking questions :)

Reply:
Ok awesome, I re-approved. :shipit: